File: greedy/removeCoveredIntervals.java
package greedy;

// https://leetcode.com/problems/remove-covered-intervals/


import java.util.Arrays;
public class removeCoveredIntervals {
    public static void main(String[] args) {
    
        SolutionRCI solution= new SolutionRCI ();
        System.out.println(solution.removeCoveredIntervals(new int [][]{{1,2},{1,4},{3,8}}));  
    }
}

class SolutionRCI {

    // Submitted by @ Jiganesh
    // TC : O(NlogN)
    // SC : O(1)

    // Runtime: 5 ms, faster than 66.26% of Java online submissions for Remove Covered Intervals.
    // Memory Usage: 39.3 MB, less than 70.35% of Java online submissions for Remove Covered Intervals.

    public int removeCoveredIntervals(int[][] intervals) {
        
        int count =intervals.length;
        Arrays.sort(intervals, (a,b)-> Integer.compare(a[0],b[0]));
        
        int c = intervals[0][0];
        int d = intervals [0][1];
        
        for (int i =1 ; i< intervals.length ;i++){
            int a = intervals[i][0];
            int b = intervals[i][1];
            if ((c<=a  && b<=d)|| (a<=c && d<=b)) {
                count--;
                d = (b>d)?b:d;
            }
            else {
                c=a;
                d=b;
            }
        }
        return count;   
    }
}

File: greedy/removeCoveredIntervals.py
# https://leetcode.com/problems/remove-covered-intervals/

class Solution:    
    
    # Submitted by @Jiganesh
    
    # Runtime: 68 ms, faster than 98.46% of Python online submissions for Remove Covered Intervals.
    # Memory Usage: 14.1 MB, less than 15.38% of Python online submissions for Remove Covered Intervals.


    # TC : O(NLogN)
    # SC : O(1)
    
    # literally merge interval and keep track
    def removeCoveredIntervals(self, intervals):
        intervals.sort(key = lambda x: x[0])
        
        c= intervals[0][0]
        d = intervals[0][1]
        count = len(intervals)
        
        for i in range ( 1, len(intervals)):
            a = intervals [i][0]
            b = intervals [i][1]
            
            if (c<=a and b<=d ) or (a<=c and d<=b):
                count-=1
                d = max(b,d)
            else:
                c= intervals[i][0]
                d = intervals[i][1]
        return count
    
    # TC : O(NLogN)
    # SC : O(1)
    # Smart , Intuative and Aesthetic solution
    
    # If merge possible - reduce that block 
    
    def removeCoveredIntervalsApproach2(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        
        intervals.sort(key = lambda x: (x[0],-x[1]))
        count = len(intervals)
        
        prev = intervals[0][1]
        for i in range(1, len(intervals)):
            if intervals[i][1] <= prev: count-=1
            else : prev = intervals [i][1]
                
        return count;
    